server/bundler/babel/babel-plugin-transform-wpcalypso-async/test/index.js

```diff
@@ -36,9 +36,9 @@ describe( 'babel-plugin-transform-wpcalypso-async', () => {
 				const code = transform( 'export default () => <AsyncLoad require="foo" />;' );
 
 				expect( code ).toBe(
-					'var _ref = function (callback) {\n  require.ensure("foo", function (require) {\n    ' +
-						'callback(require("foo").__esModule ? require("foo").default : require("foo"));\n  ' +
-						'}, "async-load-foo");\n};\n\nexport default (() => <AsyncLoad require={_ref} />);'
+					'var _ref = function (callback) {\n  import(\n  /*webpackChunkName: "async-load-foo"*/\n  "foo").then(function load(mod) {\n' +
+						'    callback(mod.__esModule ? mod.default : mod);\n' +
```

Comment:
I'm wondering if this `__esModule` dance is really necessary with the `import()` statement. I would expect webpack itself to do the interop work in case a CommonJS module is imported.

Reply:
Out of curiosity, I did some weekend webpack source reading and figured out that webpack indeed does the interop stuff for us:
- when parsing the module source, webpack [detects the module type](https://github.com/webpack/webpack/blob/2361995a429794d372d61e3fb628dea3929e0ca4/lib/dependencies/HarmonyDetectionParserPlugin.js) (ES or CommonJS) and sets the `exportsType` flag accordingly.
- then, when generating the `__webpack_require__.t` call for a module, it uses the `exportsType` flag to [determine the `mask` bitmask](https://github.com/webpack/webpack/blob/2361995a429794d372d61e3fb628dea3929e0ca4/lib/RuntimeTemplate.js#L109-L122) parameter
- finally, the runtime implementation of `__webpack_require__.t` uses the `mask` parameter to either [convert the module exports to ES format or not](https://github.com/webpack/webpack/blob/2361995a429794d372d61e3fb628dea3929e0ca4/lib/MainTemplate.js#L302-L323).

Babel has similar functionality (`interopRequire*` helpers) when transpiling `import` statements into `require`s and when transpiling `export` statements into CommonJS `export`s. But we don't use that in Calypso -- we tell Babel to leave the `import` and `export` statements and they are and leave them to webpack to handle.

Reply:
Interesting. So we really only need the branch that passed `.default`, since it'll always be there, even for CommonJS modules. And if we wanted to alter the interface, we could pass the whole module object down, allowing users to pick which export they really want to use. Sound right?

Reply:
> we really only need the branch that passed .default, since it'll always be there, even for CommonJS modules.

There's one more gotcha: the `.default` will always be there only for `import` calls, both static and dynamic, but not for `require`. Here's an example that imports an ESM and CJS module in three ways and each time logs the default exports (i.e., the `{esm|cjs}-default` strings):

_index.js_
```js
const rm = require("./module");
const rc = require("./common");

console.log("r-module", rm.default);
console.log("r-common", rc);

import im from "./module";
import ic from "./common";

console.log("i-module", im);
console.log("i-common", ic);

import("./module").then(dm => {
  import("./common").then(cm => {
    console.log("d-module", dm.default);
    console.log("d-common", cm.default);
  });
});
```

_module.js_
```js
export default 'esm-default';
```

_common.js_
```js
module.exports = 'cjs-default';
```

Note how when `import`-ing, the module type doesn't matter, but when `require`-ing, it does. `require( 'cjs-module' )` doesn't add the ESM interop. The `m.__esModule ? m.default : m` is therefore still needed.

Reply:
so then we only really need it for the sync `require` branch?

Reply:
Yes, only for the sync `require`. `import` statements do it automatically.